### PR TITLE
[cdac] Implement a JSON contract reader

### PR DESCRIPTION
**Pull request by lambdageek**
Contributes to #99298 

Implement a parser for the "compact" JSON contract descriptor format specified in [data_contracts.md](https://github.com/dotnet/runtime/blob/main/docs/design/datacontracts/data_descriptor.md)

The data model is a WIP - it's likely we will want something a bit more abstract (and less mutable).

This is not wired up to consume a real contract descriptor from target process memory at the moment.  It's only exercised by unit tests for now.

TODO:

- [x] hook up the unit tests to some testing infrastructure - either `src/libraries` or `src/tests` or to something else

### CODE REVIEW

File: src/coreclr/tools/cdac-build-tool/Program.cs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+using System.Runtime.Serialization.Json;
+using System.Text;
+using System.Text.Json;
+using System.Text.Encodings.Web;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (am11) [inline src/coreclr/tools/cdac-build-tool/Program.cs:26]: nit: delete unused usings on C# sources to reduce the noise.

Comment (lambdageek) [reply src/coreclr/tools/cdac-build-tool/Program.cs:26]: @am11 thanks.  this change is actually part of #100650 (PTAL if you're interested) - and I already changed that one to remove the unused usings.  I'll rebase this current PR so it doesn't include changes from that one.  I thought I might end up sharing code, but it didn't seem useful yet.

File: src/native/managed/cdacreader/src/ContractDescriptorParser.cs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.Diagnostics.DataContractReader;
+public partial class ContractDescriptorParser
+{
+    public const string TypeDescriptorSizeSigil = "!";
```
Comment (AaronRobinsonMSFT) [inline src/native/managed/cdacreader/src/ContractDescriptorParser.cs:13]: If this is part of the specification, can we get a link?

File: src/native/managed/cdacreader/src/ContractDescriptorParser.cs [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Microsoft.Diagnostics.DataContractReader;
+
+/// <summary>
+///   A parser for the JSON representation of a contract descriptor.
+/// </summary>
+/// <remarks>
+/// See docs/design/datacontracts/data_descriptor.md for the format
```
Comment (am11) [inline src/native/managed/cdacreader/src/ContractDescriptorParser.cs:16]: nit:
```suggestion
/// <see href="https://github.com/dotnet/runtime/blob/main/docs/design/datacontracts/data_descriptor.md">See design doc</see> for the format.
```

**Comment by lambdageek** [comment]
By the way, the unit tests for cdacreader aren't hooked up to any CI infrastructure at the moment.  I'm not sure if we should include them into `src/tests` or `src/libraries` or directly into some yaml pipeline.  Open to suggestions.

**Comment by jkoritzinsky** [comment]
For hooking up the unit tests, I think hooking it up in the same way as the ILCompiler and ILLink unit tests will be the easiest way.

That way it won't take on additional dependencies or complexity introduced by the other test harnesses.

File: src/native/managed/cdacreader/tests/Directory.Build.props [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<Project>
+  <Import Project="..\..\..\..\libraries\Directory.Build.props" />
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/native/managed/cdacreader/tests/Directory.Build.props:2]: `dotnet.sh build ... /t:Test` doesnt' seem to work without _something_ from libraries.   It's probably `eng/testing/tests.props`, but I'm not sure.  I'm cargo-culting from how the FunctionalTests are set up.

**Comment by lambdageek** [comment]
> For hooking up the unit tests, I think hooking it up in the same way as the ILCompiler and ILLink unit tests will be the easiest way.

@jkoritzinsky just so I'm clear: I would need to add a new lane in the yaml file, right?

File: src/native/managed/cdacreader/src/ContractDescriptorParser.cs [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+        public override void Write(Utf8JsonWriter writer, TypeDescriptor value, JsonSerializerOptions options)
+        {
+            throw new NotImplementedException();
+        }
+    }
+
+    internal sealed class FieldDescriptorConverter : JsonConverter<FieldDescriptor>
+    {
+        // Compact Field descriptors are either one or two element arrays
+        // 1. [number] - no type, offset is given as the number
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/native/managed/cdacreader/src/ContractDescriptorParser.cs:140]: Oops. You're right. There are just two cases: `number` or `[number, string]`. The converter also handled `[number]` which is not valid.  I'm removing it.

**Comment by jkoritzinsky** [comment]
You could add a subset and add that subset to the CLR Tools Tests job's build args (and update the path triggers). Then it will run as part of that job. No need for an additional job.

**Comment by lambdageek** [comment]
Added a `tools.cdacreadertests` subset - added it to the `CLR_Tools_Tests` test job

File: eng/Subsets.props [hunk 6]
```diff
@@ -369,6 +369,10 @@
       Test="true" Category="clr" Condition="'$(DotNetBuildSourceOnly)' != 'true' and '$(NativeAotSupported)' == 'true'"/>
   </ItemGroup>
 
+  <ItemGroup Condition="$(_subset.Contains('+tools.cdacreadertests+'))">
```
Comment (jkoritzinsky) [inline eng/Subsets.props:372]: You also need to declare the subset above with a `SubsetName` item (that will fix the validation error and also provide some basic docs for the subset)

**Review by elinor-fung — COMMENTED** [review]
Are there docs anywhere about tools tests (adding, running, etc) that we should update?

**Comment by lambdageek** [comment]
> Are there docs anywhere about tools tests (adding, running, etc) that we should update?

Added a new PR with docs: https://github.com/dotnet/runtime/pull/101186

**Comment by lambdageek** [comment]
> ```
> .dotnet/sdk/9.0.100-preview.3.24204.13/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(266,5): error NETSDK1047: (NETCORE_ENGINEERING_TELEMETRY=Build) Assets file '/__w/1/s/artifacts/obj/cdacreader/project.assets.json' doesn't have a target for 'net9.0/linux-x64'. Ensure that restore has run and that you have included 'net9.0' in the TargetFrameworks for your project. You may also need to include 'linux-x64' in your project's RuntimeIdentifiers.
> ```

there's a related  failure on the "CLR_Tools_Tests" lane that is similar to what we previously saw with the `src/native/managed` infrastructure - it's something about the right properties not flowing down to the Restore target, I think.

**Review by elinor-fung — APPROVED** [review]
Looks good to me - assuming whatever is going on with the CLR_Tools_Tests leg is straightforward to address.

**Comment by lambdageek** [comment]
Ah. The issue was that we now have two `<ProjectReference Include="cdacreader.csproj">` items.  One in compile-native.proj and the other in the unit test.  But they had distinct `AdditionalProperties` metadata, and the one in `compile-native.proj` set `Configuration` and `RuntimeIdentifier`, whereas the test didn't.  as a result the restore created a project.assets.json that was only correct for one reference and not the other.  the fix is to set the same `AdditionalProperties` on both references.

We set `RuntimeIdentifier=$(OutputRID)` so that in the case of cross-arch build NativeAOT will target the target architecture.  but I wonder if there's a better way to do that than by messing with ProjectReference - which is apparently kind of fragile.

File: src/native/managed/subproject.props [hunk 7]
```diff
@@ -0,0 +1,19 @@
+<Project>
+    <!-- properties to pass down to the subproject builds -->
+    <ItemGroup>
+        <SubprojectProps Include="Configuration" Value="$(Configuration)" />
+        <SubprojectProps Include="RuntimeConfiguration" Value="$(RuntimeConfiguration)" />
+        <SubprojectProps Include="LibrariesConfiguration" Value="$(LibrariesConfiguration)" />
+        <SubprojectProps Include="RuntimeIdentifier" Value="$(OutputRID)" />
+
+        <SubprojectProps Include="NativeLib" Value="$(NativeLibKind)" />
```
[the remark this answers is not in the document]
Comment (lambdageek) [reply src/native/managed/subproject.props]: Good catch. there are a couple of these `AdditionalProperties` that are only relevant when we're doing a proper native compilation.  I moved those back to compile-native.proj

**Pull request merged**